### PR TITLE
[MBL-1194] Implement Validate Checkout 

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

When tapping pledge, we need to run this mutation so the backend can do some extra validations before we ask Stripe to do its thing.

**For New Payment Methods**
- We'll use the payment intent that is passed back to the VM via PaymentSourceSelected
- We can then call Stripe's STPAPIClient.shared.retrievePaymentIntent to get the required paymentSourceId (paymentMethodId in Stripe)

**For Pre-Existing Payment Methods that were originally created with a setup intent**
- We'll need to call Create Payment Intent again to pass to this mutation
- We'll also need to use the selected existing card's id to filter the user's stored cards and grab the `stripeCardId`

I opted to make separate calls to this mutation depending on the case since they require enough distinct information to warrant it. There might be an opportunity to remove some duplicate code so if there are any nits please call them out!

### The mutation accepts the following params:
- the **checkoutId** (from the https://github.com/kickstarter/ios-oss/pull/1982 on ConfirmDetailsViewController). One update to this was made. Validate Checkout requires that the checkoutId be decoded to base 64.
- the **paymentMethodId** that we can pull in using, `STPAPIClient.shared.retrievePaymentIntent(withClientSecret: clientSecret)` for new cards, or the `stripeCardId` property on existing cards.
- and the **paymentIntentClientSecret** that we create using the backend.

### Extras
I also updated PostCampaignCheckoutViewController to use our `MessageBannerViewControllerPresenting` to show error messages when an API call fails.
   - It's important to note that in the case of a failure, the expected UX is to navigate the backer back to the ConfirmDetails screen so that we can being the process of creating a checkout Id/validating/etc. before attempting another pledge. This is because once we confirm the payment with Stripe, the backend will need to invalidate that request, essentially processing a refund to the backer and invalidating our previous validations.

# 🤔 Why

Gotta validate, ya dig?

# ✅ Acceptance criteria

- [x] Tapping Pledge, after adding a card or selecting an existing card, successfully runs the validateCheckout mutation with the correct values.

# ⏰ TODO

- [ ] Next step is to confirm the payment intent via Stripe.confirmPayment()

### CODE REVIEW

File: Library/ViewModels/ConfirmDetailsViewModel.swift [hunk 1]
```diff
@@ -353,7 +353,16 @@ public class ConfirmDetailsViewModel: ConfirmDetailsViewModelType, ConfirmDetail
...
+      .map { values in
+        var checkoutId: String?
+
+        if let decoded = decodeBase64(values.checkout.id), let range = decoded.range(of: "Checkout-") {
+          let id = decoded[range.upperBound...]
+          checkoutId = String(id)
+        }
+
+        return checkoutId
+      }
```
Comment (scottkicks) [inline Library/ViewModels/ConfirmDetailsViewModel.swift:365]: These will always decode to include "Checkout-", but I'm not a big fan of hardcoding strings like this

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 2]
```diff
@@ -253,11 +272,61 @@ extension PostCampaignCheckoutViewController: PledgeViewCTAContainerViewDelegate
...
+// MARK: - PledgePaymentMethodsViewControllerDelegate
+
+extension PostCampaignCheckoutViewController: PledgePaymentMethodsViewControllerDelegate {
+  func pledgePaymentMethodsViewController(
+    _: PledgePaymentMethodsViewController,
+    didSelectCreditCard paymentSource: PaymentSourceSelected
+  ) {
+    switch paymentSource {
+    case let .paymentIntentClientSecret(clientSecret):
+      return STPAPIClient.shared.retrievePaymentIntent(withClientSecret: clientSecret) { intent, _ in
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:293]: It would follow our existing patterns better if this call to Stripe's API went into `PostCampaignPledgeViewModel`, instead - that way, all the business logic stays in the VM, including the call back out to the error banner.

Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:293]: 🤔 Unless we explicitly keep stripe stuff out of VMs so that the VMs are more testable. Could go either way on this one, now that I say it.

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 3]
```diff
@@ -253,11 +272,61 @@ extension PostCampaignCheckoutViewController: PledgeViewCTAContainerViewDelegate
...
+        guard let intent = intent else {
+          self.messageBannerViewController?
+            .showBanner(with: .error, message: Strings.Something_went_wrong_please_try_again())
+          return
+        }
+
+        let stripeCardId = intent.stripeId
+        let paymentIntentClientSecret = paymentSource.paymentIntentClientSecret ?? ""
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:301]: This can safely be `paymentSource.paymentIntentClientSecret!` because it should always exist in this case.

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 4]
```diff
@@ -253,11 +272,61 @@ extension PostCampaignCheckoutViewController: PledgeViewCTAContainerViewDelegate
...
+
+        self.viewModel.inputs.creditCardSelected(with: (stripeCardId, paymentIntentClientSecret))
+      }
+    default:
+      break
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:306]: This also needs to handle the `.savedCreditCard` case - otherwise, we're not completing the pledge if you re-use an old card.

File: Library/ViewModels/ConfirmDetailsViewModel.swift (hunk 1 above)
Comment (amy-at-kickstarter) [reply Library/ViewModels/ConfirmDetailsViewModel.swift:365]: Huh, is this decoding a GraphID sent back from GraphQL? You should ask the server-side folks to give you a plain ID in the mutation. Those Base64 IDs are sort of intended to be opaque and shouldn't need to be read like this.

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 5]
```diff
@@ -130,6 +136,35 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+    let validateCheckout = Signal.combineLatest(checkoutId, stripeCardIdAndPaymentIntent)
+      .takeWhen(self.submitButtonTappedProperty.signal)
+      .switchMap { checkoutId, stripeCardIdAndPaymentIntent in
+        let (stripeCardId, paymentIntent) = stripeCardIdAndPaymentIntent
+
+        return AppEnvironment.current.apiService
+          .validateCheckout(
+            checkoutId: checkoutId,
+            paymentSourceId: stripeCardId,
+            paymentIntentClientSecret: paymentIntent
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:153]: Do we need a payment intent to validate? What if you're using a saved card?

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 6]
```diff
@@ -130,6 +136,35 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+          )
+          .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+          .materialize()
+      }
+
+    self.validateCheckoutSuccess = stripeCardIdAndPaymentIntent
+      .takeWhen(validateCheckout.values().ignoreValues())
+      .map { stripeCardIdAndPaymentIntent in
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:161]: Nit: Can be slightly more compact like
`.map { stripeCardId, paymentIntent  in paymentIntent }`

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 7]
```diff
@@ -130,6 +136,35 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+        let (_, paymentIntent) = stripeCardIdAndPaymentIntent
+        return paymentIntent
+      }
+
+    self.showErrorBannerWithMessage = validateCheckout.errors()
+      .map { _ in Strings.Something_went_wrong_please_try_again() }
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:167]: Nit: Will we show any of the errors to the user, or always this one?

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift (hunk 4 above)
Comment (scottkicks) [reply Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:306]: Sheesh, thanks for the call-out! I got a little lost in the sauce with this one, and save cards slipped my mind. 

For saved cards I think we'll need to grab the stripeCardId on the selected stored card in the VM to pass to this mutation.

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift (hunk 7 above)
Comment (scottkicks) [reply Library/ViewModels/PostCampaignCheckoutViewModel.swift:167]: Unfortunately, we don't have any neat server-side error messaging. I was told generic messaging is fine for now 😢

**Comment by scottkicks** [comment]
Looking into your other comments.

> Why do we need to decode the `GraphID` of a Checkout? Can this be resolved at the mutation level, instead of doing it in the client?

I spoke with the backend, and they would rather we did this. It sounded like they could (they do in other cases), but they aren't willing to spend the time updating the different pieces involved right now.

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift (hunk 2 above)
Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:294]: Sounds good, OK.

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift (hunk 4 above)
Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:307]: Yeah, I think that's an argument for updating the enum to be 
```
case .savedCreditCard(String, String)
```

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 8]
```diff
@@ -140,26 +139,53 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
         return AppEnvironment.current.apiService
           .validateCheckout(
             checkoutId: checkoutId,
             paymentSourceId: stripeCardId,
-            paymentIntentClientSecret: paymentIntent
+            paymentIntentClientSecret: paymentIntent ?? ""
           )
           .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
           .materialize()
       }
```
Comment (scottkicks) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:180]: still working on verifying what the expected value is for paymentIntentClientSecret for existing payment methods

File: Library/ViewModels/ConfirmDetailsViewModel.swift [hunk 9]
```diff
@@ -211,7 +211,7 @@ public class ConfirmDetailsViewModel: ConfirmDetailsViewModelType, ConfirmDetail
 
     /// Hide when there is a reward and shipping is enabled (accounts for digital rewards), and in a pledge context
     self.pledgeSummaryViewHidden = Signal.zip(baseReward, context).map { baseReward, context in
-      (baseReward.isNoReward == false && baseReward.shipping.enabled) && context == .pledge
+      baseReward.isNoReward == false && context == .pledge
```
Comment (scottkicks) [inline Library/ViewModels/ConfirmDetailsViewModel.swift:214]: unrelated fix for showing the summary view on details

File: Kickstarter-iOS/Features/ConfirmDetails/Controllers/ConfirmDetailsViewController.swift
Comment (scottkicks) [inline Kickstarter-iOS/Features/ConfirmDetails/Controllers/ConfirmDetailsViewController.swift:1]: unrelated cleanup of details screens. makes sure table view expands to the full width of the view

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 10]
```diff
@@ -199,6 +205,28 @@ final class PostCampaignCheckoutViewController: UIViewController {
...
+
+        // TODO: Confirm paymentIntent using Stripe.confirmPayment()
+      }
+
+    self.viewModel.outputs.validateCheckoutExistingCardSuccess
+      .observeForControllerAction()
+      .observeValues { [weak self] _ in
+        guard let self else { return }
+
+        // TODO: Confirm paymentIntent using Stripe.confirmPayment()
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:221]: Is this ``// TODO:`` correct for the existing card case, too?

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 11]
```diff
@@ -253,11 +281,70 @@ extension PostCampaignCheckoutViewController: PledgeViewCTAContainerViewDelegate
...
+    switch paymentSource {
+    case let .paymentIntentClientSecret(clientSecret):
+      return STPAPIClient.shared.retrievePaymentIntent(withClientSecret: clientSecret) { intent, _ in
+        guard let intent = intent else {
+          self.messageBannerViewController?
+            .showBanner(with: .error, message: Strings.Something_went_wrong_please_try_again())
+          return
+        }
+
+        let paymentMethodId = intent.paymentMethodId!
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:309]: Nit: Could also throw this `let` into the above `guard` statement, just in case.

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift [hunk 12]
```diff
@@ -253,11 +281,70 @@ extension PostCampaignCheckoutViewController: PledgeViewCTAContainerViewDelegate
...
+        self.viewModel.inputs
+          .creditCardSelected(
+            source: paymentSource,
+            paymentMethodId: paymentMethodId,
+            isNewPaymentMethod: true
+          )
+      }
+    case let .savedCreditCard(savedCardId):
+      self.viewModel.inputs
+        .creditCardSelected(source: paymentSource, paymentMethodId: savedCardId, isNewPaymentMethod: false)
```
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:320]: 👍

File: KsApi/models/graphql/UserCreditCards.swift [hunk 13]
```diff
@@ -9,6 +9,7 @@ public struct UserCreditCards: Decodable {
     public var id: String
     public var lastFour: String
     public var type: CreditCardType?
+    public var stripeCardId: String?
```
Comment (amy-at-kickstarter) [inline KsApi/models/graphql/UserCreditCards.swift:12]: Ah, yeah, I only added it to the GraphQL object. Good find.

File: Library/ViewModels/ConfirmDetailsViewModel.swift (hunk 9 above)
Comment (amy-at-kickstarter) [reply Library/ViewModels/ConfirmDetailsViewModel.swift:214]: 👍

File: Library/ViewModels/ConfirmDetailsViewModel.swift [hunk 14]
```diff
@@ -353,7 +353,16 @@ public class ConfirmDetailsViewModel: ConfirmDetailsViewModelType, ConfirmDetail
...
+      .map { values in
+        var checkoutId = values.checkout.id
+
+        if let decoded = decodeBase64(checkoutId), let range = decoded.range(of: "Checkout-") {
+          let id = decoded[range.upperBound...]
+          checkoutId = String(id)
+        }
+
+        return checkoutId
+      }
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/ConfirmDetailsViewModel.swift:365]: Any resolution on why this needs to be decoded from GraphQL? At a minimum, this should really have a ``//TODO: `` explaining the hack.

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 15]
```diff
@@ -130,6 +137,99 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+        AppEnvironment.current.apiService
+          .fetchGraphUser(withStoredCards: true)
+          .ksr_debounce(.seconds(1), on: AppEnvironment.current.scheduler)
+          .map { envelope in (envelope, false) }
+          .prefix(value: (nil, true))
+          .materialize()
+      }
+
+    let storedCardsValues = storedCardsEvent.values()
+      .filter(second >>> isFalse)
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:157]: What does this do?

Comment (amy-at-kickstarter) [reply Library/ViewModels/PostCampaignCheckoutViewModel.swift:157]: Nit: Would prefer this to be a non-Prelude operator for clarity.

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 16]
```diff
@@ -130,6 +137,99 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+      .map(first)
+      .skipNil()
+      .map { $0.me.storedCards.storedCards }
+
+    let selectedExistingCard = self.creditCardSelectedProperty.signal.skipNil()
+      .filter { $0.isNewPaymentMethod == false }
+
+    let newPaymentIntentEvent = initialData
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:165]: This only needs to happen when you're _not_ adding a new card via the payment method selector, since in that case, we've already made a PI for you. Should this be attached to a different event than `initialData`?

File: Library/ViewModels/PostCampaignCheckoutViewModel.swift [hunk 17]
```diff
@@ -130,6 +137,99 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
...
+    self.validateCheckoutExistingCardSuccess = paymentIntentClientSecret
+      .takeWhen(validateCheckoutExistingCard.values())
+      .map { $0 }
+
+    // MARK: - Validate New Cards
+
+    let selectedNewCreditCard = self.creditCardSelectedProperty.signal.skipNil()
+      .filter { $0.isNewPaymentMethod }
+
+    // Runs validation for new cards that were created with payment intents.
```
Comment (amy-at-kickstarter) [inline Library/ViewModels/PostCampaignCheckoutViewModel.swift:210]: Loving these comments

**Review by amy-at-kickstarter — APPROVED** [review]
LGTM with some questions and nits

File: Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift (hunk 10 above)
Comment (scottkicks) [reply Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift:221]: Yes we'll have to confirm the payment either way using the intent

File: Library/ViewModels/ConfirmDetailsViewModel.swift (hunk 14 above)
Comment (scottkicks) [reply Library/ViewModels/ConfirmDetailsViewModel.swift:365]: Ah yea i left an unfortunate [comment](https://github.com/kickstarter/ios-oss/pull/1999#issuecomment-2020596093) earlier.

Comment (scottkicks) [reply Library/ViewModels/ConfirmDetailsViewModel.swift:365]: its silly that they would give us a string, that we only use to send to one of their mutations, and not format it correctly for us
